WIP: API documentation and GLM introduction (port from notebook)

titipata opened the pull request:
- API documentation page
- Tutorial page (work-in-progress), I move the tutorial notebook to the page since it's more like a reading page rather than coding notebook.

Todos before merging
- [x] Use auto-doc extension
- [x] Use `.py` files for documentation instead of `.rst` files so that they are executable
- [x] delete notebooks folder

jasmainak commented on doc/API-documentation.rst:
```diff
@@ -0,0 +1,104 @@
+API Documentation
+=================
+
+:class:`GLM` Objects
```
woops, there is a sphinx extension for this so you don't have to do the hard work ;) 

sorry I didn't realize you would jump into it! Here you go: http://www.sphinx-doc.org/en/stable/ext/autodoc.html

titipata replied:
Oh boy, I didn't know that we can use autodoc! I was about to ask.

jasmainak commented on doc/GLM-introduction.rst:
```diff
@@ -0,0 +1,435 @@
+**Poisson-like GLM**
+
+The ``pyglmnet`` implementation comes with `poisson`, `poissonexp`, `binomial`,
+`normal` and `multinomial` distributions, but for illustration, I will walk you
```
should we change the `I` to `we` in the documentation? @pavanramkumar okay with it?

pavanramkumar replied:
sure, no problem :)

jasmainak commented on doc/GLM-introduction.rst:
```diff
@@ -0,0 +1,435 @@
+For numerical reasons, let's adopt a stabilizing non-linearity, known as the
+softplus or the smooth rectifier `Dugas et al., 2001 <http://papers.nips.cc/paper/1920-incorporating-second-order-functional-knowledge-for-better-option-pricing.pdf>`_,
+and adopted by Jonathan Pillow's and Liam Paninski's groups for neural data analysis.
+See for instance: `Park et al., 2014 <http://www.nature.com/neuro/journal/v17/n10/abs/nn.3800.html>`_.
```
if you want the code to be executable, you should put it in the `examples/` folder ... take a look at the two files I put there. Are you trying to make something which is complementary to the two ...

jasmainak replied:
basically `sphinx-gallery` does the hard work of converting the `.py` file into `.rst` as well as `.ipynb` ...

jasmainak commented:
maybe another tip, if you want to not send coverage messages to everyone -- you can add `[ci skip]` in your commit message. Only in the last message, don't put it so that we are sure Travis is happy before merging. That way you skip travis tests for all but your last commit -- and it's more environment friendly too :)

jasmainak commented:
@titipata I added some Todos to the project description by editing your first comment !

pavanramkumar commented:
wow the autodoc feature seems powerful! as usual, learning a ton! thanks for the hard work @titipata @jasmainak

jasmainak commented:
@titipata you may need to add it to `conf.py` to make it work ...

pavanramkumar commented:
ok, circleci works! https://circle-artifacts.com/gh/pavanramkumar/pyglmnet/2/artifacts/0/home/ubuntu/pyglmnet/doc/_build/html/index.html

jasmainak commented:
indeed that's the idea ;)

you should see it in the PRs from now on ! if there is a problem in the docs, circleCI should fail (we should check this too by making a commit where there is a problem in the docs) ...

titipata commented:
@jasmainak, my commits are a pretty messy. However, here is the summary

- I moved example to `plot_glmnet.py`. This page is more like a tutorial about GLM so I wish that it could be a tab that people can click from the first page i.e. `introduction to GLM` >> `plot_glmnet.py`
- I tried `auto-doc extension`, however, it still does not show the link to GLM. Maybe you can help me fix that problem (thanks in advance). We could have everything in one page since we have only one big class, I guess. (like this page: http://kbarbary.github.io/nestle/api.html)
- I added `developers.rst` so everything will be in one place.

jasmainak commented:
@titipata I'll give it a shot after lunch :) don't touch the branch. I'll push to this branch itself ...

jasmainak commented on doc/conf.py:
```diff
@@ -14,11 +14,12 @@
-#sys.path.insert(0, os.path.abspath('.'))
+
+# import pyglmnet
+sys.path.insert(0, os.path.abspath('..'))
```
hmm ... you shouldn't have to do this. Did you install `pyglmnet` ?

jasmainak commented on pyglmnet/__init__.py:
```diff
@@ -1 +1,2 @@
 from .pyglmnet import GLM
+__version__ = '0.1.dev'
```
cool :)

jasmainak commented on doc/Makefile:
```diff
@@ -42,7 +42,7 @@ help:
 	rm -rf $(BUILDDIR)/*
 	rm -rf auto_examples/
-	rm -rf modules/generated/*
+	rm -rf generated/*
```
@titipata this was the culprit. You owe me a beer for this one ;)

titipata replied:
Hahaaaaa, prefect catch. Thanks @jasmainak. I'll continue the rest.

jasmainak commented:
I'll let you continue on the rest of the documentation @titipata

jasmainak commented:
https://circle-artifacts.com/gh/pavanramkumar/pyglmnet/9/artifacts/0/home/ubuntu/pyglmnet/doc/_build/html/generated/pyglmnet.GLM.html#pyglmnet.GLM for your reference :)

jasmainak commented:
also maybe a minor cosmit but I noticed that it's really hard to tell if something has a link or not in the API reference in the `read_the_docs` template. Maybe we can modify the css and make it more obvious ...

titipata commented:
@pavanramkumar and @jasmainak, I think it looks pretty good now. Can you both check before we accept PR?

jasmainak commented:
@titpata, it looks like this example is broken: https://circle-artifacts.com/gh/pavanramkumar/pyglmnet/11/artifacts/0/home/ubuntu/pyglmnet/doc/_build/html/auto_examples/plot_poisson.html#sphx-glr-auto-examples-plot-poisson-py

circleci in action @hugoguh ;)

jasmainak commented:
Great work @titipata !

jasmainak commented:
should we also add the examples gallery to the contents, so it's always in the navigation bar?

jasmainak commented:
looks like you need to rebase with respect to master ... I think that's why circle CI is failing.

titipata commented:
@jasmainak, yeah, we should add examples to the content. Either a link 'Gallery' on the tab or on first page are good.

jasmainak commented:
+1 for merge from my side once Travis and Circle are happy :)

pavanramkumar commented:
thanks so much @titipata!!

- circleci is failing because of `scikits.bootstrap`. although it is not a dependency in `pyglmnet`, we use it in the example to show how to bootstrap. fix: remove the bootstrap example, or then add `scikits.bootstrap` to requirements.txt
- travisci is failing due to randomness: we haven't perfectly solved the `numpy.random.seed` issue

titipata commented:
I remove bootstrap. @jasmainak you can now wrap up small changes and we are ready to go, I guess :)

jasmainak commented:
Thanks @titipata! For the `scikits.bootstrap` issue -- adding to requirements won't fix it I think. You'll have to edit the circleCI yml file to install the right version of `scikit-learn` ?

titipata commented:
@jasmainak Yeah, I just remove bootstrap. We can consider adding that snippet later on. I don't have to edit circleCI to right version of `scikit-learn`. We only need minor fix as mentioned, put Example on the tab. 

@pavanramkumar suggested moving [this page](http://pavanramkumar.github.io/pyglmnet/auto_examples/plot_glmnet.html#sphx-glr-auto-examples-plot-glmnet-py) to `tutorial.rst` since it aims to be tutorial page rather than gallery.

jasmainak commented:
> @pavanramkumar suggested moving this page to tutorial.rst since it aims to be tutorial page rather than gallery.

Do you need a separate folder for tutorials? You can also do this: http://sphinx-gallery.readthedocs.io/en/latest/advanced_configuration.html. That way, you will save yourself the trouble of moving it + ensure that the code actually works ...

titipata commented:
Ahh, I see. I can do that later on seems like separate folders will be suitable for this case.
